rust/foxglove/src/websocket/protocol/client.rs

```diff
@@ -182,21 +188,80 @@ pub(crate) struct Unsubscribe {
 // https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-advertise
 #[derive(Debug, Deserialize, PartialEq)]
 #[serde(rename_all = "camelCase")]
+pub(crate) struct JsonClientAdvertise {
+    pub channels: Vec<JsonClientChannel>,
+}
+
+#[derive(Debug, PartialEq)]
 pub(crate) struct ClientAdvertise {
     pub channels: Vec<ClientChannel>,
 }
 
+impl TryFrom<JsonClientAdvertise> for ClientAdvertise {
+    type Error = ParseError;
+    fn try_from(msg: JsonClientAdvertise) -> Result<Self, ParseError> {
+        Ok(ClientAdvertise {
+            channels: msg
+                .channels
+                .into_iter()
+                .map(ClientChannel::try_from)
+                .collect::<Result<_, _>>()?,
+        })
+    }
+}
+
 #[derive(Debug, Clone, Deserialize, PartialEq)]
 #[serde(rename_all = "camelCase")]
-pub(crate) struct ClientChannel {
+pub(crate) struct JsonClientChannel {
     pub id: ClientChannelId,
     pub topic: String,
     pub encoding: String,
     pub schema_name: String,
     pub schema_encoding: Option<String>,
+    /// May be base64-encoded depending on the schema_encoding.
     pub schema: Option<String>,
 }
 
+#[derive(Debug, PartialEq)]
+pub(crate) struct ClientChannel {
+    pub id: ClientChannelId,
+    pub topic: String,
+    pub encoding: String,
+    pub schema_name: String,
+    pub schema_encoding: Option<String>,
+    /// May or may not be a UTF-8 string depending on the schema_encoding.
+    pub schema: Option<Vec<u8>>,
+}
+
+impl TryFrom<JsonClientChannel> for ClientChannel {
+    type Error = ParseError;
+    fn try_from(channel: JsonClientChannel) -> Result<Self, ParseError> {
+        let schema = channel
+            .schema
+            .map(|schema| {
+                if channel
+                    .schema_encoding
+                    .as_ref()
+                    .map(|enc| is_known_binary_schema_encoding(enc.as_str()))
+                    .unwrap_or(false)
```

Comment:
Nit: Could do

```suggestion
                    .is_some_and(|enc| is_known_binary_schema_encoding(enc.as_str()))
```

You could even whittle that down to `.is_some_and(is_known_binary_schema_encoding)` if you make that function generic over `AsRef<str>`.